examples/rbfe_single.py

```diff
@@ -120,31 +199,102 @@ def wrap_method(args, fxn):
     ordered_params = forcefield.get_ordered_params()
     ordered_handles = forcefield.get_ordered_handles()
 
-    gradient_clip_thresholds = {
-        nonbonded.AM1CCCHandler: 0.05,
-        nonbonded.LennardJonesHandler: np.array([0.001, 0])
-    }
+    handle_types_being_optimized = [AM1CCCHandler, LennardJonesHandler]
+
+    def flatten(params) -> Tuple[array, callable]:
+        """Turn params dict into flat array, with an accompanying unflatten function
+
+        TODO: note that the result is going to be in the order given by ordered_handles (filtered by presence in hrandle_types)
+            rather than in the order they appear in handle_types_being_optimized
+
+        TODO: maybe leave out the reference to handle_types_being optimized altogether
+
+        TODO: does Jax have a pytree-based flatten / unflatten utility?
+        """
+
+
+        theta_list = []
+        _shapes = dict()
+        _handle_types = []
+
+        for param, handle in zip(params, ordered_handles):
+            assert handle.params.shape == param.shape
+            key = type(handle)
+
+            if key in handle_types_being_optimized:
+                theta_list.append(param.flatten())
+                _shapes[key] = param.shape
+                _handle_types.append(key)
+
+        theta = np.hstack(theta_list)
 
-    for epoch in range(100):
+        def unflatten(theta: array) -> Dict[Handler, array]:
+            params = dict()
+            i = 0
+            for key in _handle_types:
+                shape = _shapes[key]
+                num_params = int(np.prod(shape))
+                params[key] = np.array(theta[i : i + num_params]).reshape(shape)
+                i += num_params
+            return params
+
+        return theta, unflatten
+
+
+
+    # in each optimization step, don't step so far that you think you're jumping to
+    #   loss_next = relative_improvement_bound * loss_current
+    relative_improvement_bound = 0.95
+
+    flat_theta_traj = []
+    flat_grad_traj = []
+    loss_traj = []
+
+    for epoch in range(1000):
         epoch_params = serialize_handlers(ordered_handles)
         loss, loss_grad = vg_fn(ordered_params, mol_a, mol_b, core, label_ddG)
 
         print("epoch", epoch, "loss", loss)
 
-        for loss_grad, handle in zip(loss_grad, ordered_handles):
-            assert handle.params.shape == loss_grad.shape
+        # note: unflatten_grad and unflatten_theta have identical definitions for now
+        flat_loss_grad, unflatten_grad = flatten(loss_grad)
+        flat_theta, unflatten_theta = flatten(ordered_params)
+
+        step_lower_bound = loss * relative_improvement_bound
+        theta_increment = _smart_clip(flat_theta, loss, flat_loss_grad, step_lower_bound=step_lower_bound)
```

Comment:
Why does the smart_clip have to run over a single flattened view over the individual handler's .params? Is it a speed issue?

Reply:
PS if you're confident that a single step_size/lower_bound pair is sufficient across all of our handlers - then maybe this is still okay. It just seems a little unwieldy to flatten -> compute updates -> unflatten -> apply updates

Reply:
Not speed, but flexibility. (I think speed of this operation is pretty irrelevant -- the cost of essentially doing a "-=" (or anything up to the order of a few seconds of thinking to decide what to do with the current gradient estimate) is negligible compared with cost of running a free energy calculations.)

I think providing a flat view of the function improves flexibility, and our ability to try different optimization strategies. Generic optimizers typically assume access to a function in terms of queries to `f(x)`, `grad(f)(x)`, ... , where `x` is a flat array. Adding a parameter flattening / unflattening utility is intended to allow the gradient-processing logic to be more disentangled from the ff_handlers structure. The definition of the current step function `_smart_clip`, like the definition of most any optimizer we may import from another library, does not need to know anything about how we're currently organizing ff_handlers, and will not need to be refactored if we decide to update how parameters are stored in ff_handlers. (For example, if we later use a different charge model than `AM1CCC`, or represent torsion terms in a way that associates only one term to each torsion tuple.)

Reply:
Okay - we can try this approach for now then. Just be aware of our need to

1) sometimes set certain components to zero (eg. periods in torsions), which might be difficult to identify/mark in a flattened view
2) if we ever want to take an approach like forcebalance to convert between mathematical and physical parameters by attaching units. 

but either way, I'm comfortable with letting you make the call on this (with the assumption that you'll be able to figure out a good solution to the above should the need arise!)